wpsc-includes/variations.class.php

```diff
@@ -296,3 +296,117 @@ function wpsc_get_child_object_in_terms_var( $parent_id, $terms, $taxonomies, $a
+ * );
+ *
+ * Here the keys are captured when someone clicks the 'add to cart' button and correspond with ... whatever. They don't really matter for our function.
+ * Really ou could pass an array of 2 term_ids here and a $product_id that has variations to match those terms and you'd get back the expected array.
+ *
+ * @since 4.0
+ *
+ * @param array     $variations     required            The array of variation selections
+ * @param int       $product_id     required            The default product_id
+ * @uses wpsc_get_child_objects_in_terms()              Given $product_id and product params this returns the variation product id
+ * @return array    $args {
+ *      @param  int     product_id                      The variation product_id
+ *      @param  array   variation_values                The array of variation_values that wpsc_add_to_cart needs to populate $provided_parameters['variation_values']
+ * }
+ */
+function wpsc_get_product_data_from_variations( $variations, $product_id ) {
+
+	$variation_values = array();
+
+	foreach ( (array) $variations as $key => $variation ) {
+		$variation_values[ (int) $key ] = (int) $variation;
+	}
+
+	if ( count( $variation_values ) > 0 ) {
+		$variation_product_id = wpsc_get_child_object_in_terms( $product_id, $variation_values, 'wpsc-variation' );
+		if ( $variation_product_id > 0 ) {
+			$product_id = $variation_product_id;
+		}
+	}
+
+	/**
+	 * Allows users to filter the product_id based on the variation selections
+	 *
+	 * @since 4.0
+	 *
+	 * @param int   $product_id             The default passed product_id
+	 * @param array $variations             The variation selections passed to the core function
+	 */
+	$product_id = apply_filters( 'wpsc_variation_product_id', absint( $product_id ), $variations );
+
+	$return_args = array(
+		'product_id'       => absint( $product_id ),
+		'variation_values' => $variation_values,
+	);
+
+	return $return_args;
+
+}
+
+/**
+ * Wrapper for wpsc_get_product_data_from_variations that returns the variation product_id
+ *
+ * See wpsc_get_product_data_from_variations for a full description on what needs to be passed in the $variations params
+ *
+ * @since 4.0
+ *
+ * @uses wpsc_get_product_data_from_variations()                        Returns array of data pertaining to product variations
+ * @return int      $product_id                                         The product_id corresponding to the selected variation
+ */
+function wpsc_get_product_id_from_variations( $variations, $product_id ){
+
+	$values = wpsc_get_product_data_from_variations( $variations, $product_id );
+
+	/**
+	 * Allows users to filter the product_id based on the variation selections
+	 *
+	 * @since 4.0
+	 *
+	 * @param int   $product_id             The default passed product_id
+	 * @param array $variations             The variation selections passed to the core function
+	 */
+	$product_id = apply_filters( 'wpsc_variation_product_id', absint( $values['product_id'] ), $variations );
```

Comment:
Let's add `$values` and the originally passed `$product_id` as additional parameters to the filter.

And instead of setting the $product_id variable here, let's simply return it, e.g. `return absint( apply_filters( 'wpsc_variation_product_id', absint( $values['product_id'] ), $variations, $product_id, $values ) );`